packages/debug/src/browser/debug-configuration-manager.ts

```diff
@@ -153,11 +171,61 @@ export class DebugConfigurationManager {
     get current(): DebugSessionOptions | undefined {
         return this._currentOptions;
     }
+
+    async getSelectedConfiguration(): Promise<DebugSessionOptions | undefined> {
+        // providerType applies to dynamic configurations only
+        if (!this._currentOptions?.providerType) {
+            return this._currentOptions;
+        }
+
+        // Refresh a dynamic configuration from the provider,
+        // This allow providers to update properties before the execution e.g. program
+        const { providerType, configuration: { name } } = this._currentOptions;
+        const configuration = await this.fetchDynamicDebugConfiguration(name, providerType);
+
+        if (!configuration) {
+            const message = nls.localize(
+                'theia/debug/missingConfiguration',
+                "Dynamic configuration '{0}:{1}' is missing or not applicable", providerType, name);
+            throw new Error(message);
+        }
+
+        return { configuration, providerType };
+    }
+
     set current(option: DebugSessionOptions | undefined) {
         this.updateCurrent(option);
+        this.updateRecentlyUsedDynamicConfigurationOptions(option);
     }
+
+    protected updateRecentlyUsedDynamicConfigurationOptions(option: DebugSessionOptions | undefined): void {
+        if (option?.providerType) { // if it's a dynamic configuration option
+            // Removing an item already present in the list
+            const index = this.recentDynamicOptionsTracker.findIndex(item => this.dynamicOptionsMatch(item, option));
+            if (index > -1) {
+                this.recentDynamicOptionsTracker.splice(index, 1);
+            }
+            // Adding new item, most recent at the top of the list
+            if (this.recentDynamicOptionsTracker.unshift(option) > 3) {
+                // Remove oldest, i.e. Keeping a short number of recently used
+                // configuration options
+                this.recentDynamicOptionsTracker.pop();
+            }
```

Comment:
There are only two places where we add options to the tracker 
1) With the `unshift` above
2) When restoring options saved in a previous session

So practically only 1) is relevant (unless local storage is manipulated), so we are checking directly after each addition, 
since we add items to the tracker one by one, I don't see how we could go higher than the hard coded limit. 
If you could please expand on the scenario.

Reply:
@alvsan09 I just prefer more defensive coding that would also be clearer. The goal is to have 3 elements as not to clutter the dropdown, but the current code does not guarantee it, for example if in the future there are additional callers that add additional items to the list. (I believe it is why vscode uses `slice`) 

I'll leave the decision up to you however! :)

Reply:
I like the idea of using slice! As Vincent said this will ensure that the list always has the expected length.

Reply:
I agree: 
The new code uses `splice` as follows: 
                this.recentDynamicOptionsTracker.splice(recentMax);